### PR TITLE
Fix daemon-set-controller bootstrap RBAC policy

### LINKED ISSUE #62145: `daemon-set-controller` bootstrap RBAC policy does not include `get` permission for `controllerrevisions`
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug
/sig auth

**What happened**:

Testing a daemon-set spec, I deleted/recreated it multiple times. At one of the recreations, the daemon-set was not scheduled anymore, and the number of desired pods went to 0 (on a 3 nodes cluster):

```
$ kubectl get ds -n pks-infrastructure
NAME             DESIRED   CURRENT   READY     UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
nsx-node-agent   0         0         0         0            0           <none>          2m
```

The kube-apiserver stderr logs shows:

```
I0404 14:33:01.871517    5724 wrap.go:42] POST /apis/extensions/v1beta1/namespaces/pks-infrastructure/daemonsets: (7.588801ms) 201 [[kubectl/v1.9.5 (linux/amd64) kubernetes/f01a2bf] 30.0.1.3:38984]
I0404 14:33:01.880581    5724 wrap.go:42] POST /apis/apps/v1beta1/namespaces/pks-infrastructure/controllerrevisions: (7.853889ms) 409 [[kube-controller-manager/v1.9.5 (linux/amd64) kubernetes/f01a2bf/system:serviceaccount:kube-system:daemon-set-controller] 30.0.1.2:50964]
I0404 14:33:01.881572    5724 rbac.go:116] RBAC DENY: user "system:serviceaccount:kube-system:daemon-set-controller" groups ["system:serviceaccounts" "system:serviceaccounts:kube-system" "system:authenticated"] cannot "get" resource "controllerrevisions.apps" named "nsx-node-agent-8f5d66fd6" in namespace "pks-infrastructure"
I0404 14:33:01.881971    5724 wrap.go:42] GET /apis/apps/v1beta1/namespaces/pks-infrastructure/controllerrevisions/nsx-node-agent-8f5d66fd6: (633.24µs) 403 [[kube-controller-manager/v1.9.5 (linux/amd64) kubernetes/f01a2bf/system:serviceaccount:kube-system:daemon-set-controller] 30.0.1.2:50964]
```
The kube-controller stderr logs shows:

```
E0404 14:57:48.743754   10416 daemon_controller.go:265] pks-infrastructure/nsx-node-agent failed with : failed to construct revisions of DaemonSet: controllerrevisions.apps "nsx-node-agent-8f5d66fd6" is forbidden: User "system:serviceaccount:kube-system:daemon-set-controller" cannot get controllerrevisions.apps in the namespace "pks-infrastructure"
```

The (default) permissions for the `system:controller:daemon-set-controller` cluster role are:

```
$ kubectl describe  clusterroles system:controller:daemon-set-controller
Name:         system:controller:daemon-set-controller
Labels:       kubernetes.io/bootstrapping=rbac-defaults
Annotations:  rbac.authorization.kubernetes.io/autoupdate=true
PolicyRule:
  Resources                         Non-Resource URLs  Resource Names  Verbs
  ---------                         -----------------  --------------  -----
  events                            []                 []              [create patch update]
  nodes                             []                 []              [list watch]
  pods                              []                 []              [create delete list patch watch]
  pods/binding                      []                 []              [create]
  controllerrevisions.apps          []                 []              [create delete list patch update watch]
  daemonsets.apps                   []                 []              [get list watch]
  daemonsets.apps/finalizers        []                 []              [update]
  daemonsets.apps/status            []                 []              [update]
  daemonsets.extensions             []                 []              [get list watch]
  daemonsets.extensions/finalizers  []                 []              [update]
  daemonsets.extensions/status      []                 []              [update]
```

Seems to be missing a `get` permission for the `controllerrevisions.apps` resource. I edited manually the cluster-role adding the `get` permission and then the daemonset was scheduled again.

**What you expected to happen**:

The `system:controller:daemon-set-controller` cluster role default permissions should allow getting the `controllerrevisions.apps` resource.

**How to reproduce it (as minimally and precisely as possible)**:

I tried to reproduce it again, but I'm unable to find the exact pattern to reproduce it. 

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): `v1.9.2`
- Cloud provider or hardware configuration: `vSphere`
- OS (e.g. from /etc/os-release): 
```
NAME="Ubuntu"
VERSION="14.04.5 LTS, Trusty Tahr"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 14.04.5 LTS"
VERSION_ID="14.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
```
- Kernel (e.g. `uname -a`): `Linux dc9711b4-9208-444b-9eac-78048ca082a5 4.4.0-116-generic #140~14.04.1-Ubuntu SMP Fri Feb 16 09:25:20 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by frodenas**
**What this PR does / why we need it**: 

Adds a `get` permission for the `controllerrevisions` resource to the `daemon-set-controller` cluster role.

**Which issue(s) this PR fixes** :

Fixes #62145

**Special notes for your reviewer**:

The daemon-sets controller [constructs the history](https://github.com/kubernetes/kubernetes/blob/release-1.9/pkg/controller/daemon/daemon_controller.go#L1116) of the daemon-set, so it needs to [check the controller revisions](https://github.com/kubernetes/kubernetes/blob/release-1.9/pkg/controller/daemon/update.go#L265) for the daemon-set app. See issue above.

**Release note**:

```release-note
Resolves forbidden error when the `daemon-set-controller` cluster role access `controllerrevisions` resources.
```

### CODE REVIEW

**Comment by ericchiang** [comment]
/lgtm

Is this newly introduced behavior or will it require a cherry pick? Based on the bug report it seems like this would be extremely disruptive bug. Surprised this only surfaced recently.

**Comment by frodenas** [comment]
@ericchiang It will require a cherry-pick, at least for 1.9. I ran into this randomly, I'm still unsure what scenarios trigger the issue.

**Comment by liggitt** [comment]
Can you add a test that exercises the function that fails without this permission?

Are there other controllers that need similar permissions/tests?

**Comment by liggitt** [comment]
/assign @kow3ns 
can you give feedback on what would be needed to exercise this in CI?

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
/assign @deads2k
for approval

**Comment by deads2k** [comment]
> /assign @deads2k
> for approval

@janetkuo @liggitt (also has approval authority) is already in this thread and asked for more information/tests.

**Comment by liggitt** [comment]
still pending a test that covers this use case

**Comment by janetkuo** [comment]
@liggitt 

https://github.com/kubernetes/kubernetes/blob/31d22870b2cf0326384fab154a8adf5155ab6975/pkg/controller/daemon/update.go#L336

DaemonSet controller only calls GET ControllerRevisions when there's ControllerRevision name collision  when DaemonSet tries to create a new ControllerRevision. This code path is only called when the ControllerRevision cache store is stale or due to hash collisions. 

Is RBAC unit test not enough? We can add an integration test or e2e test to have a DaemonSet with a ControllerRevision who doesn't match current DaemonSet template but has name collision with the ControllerRevision that the DaemonSet is going to create. 

It's preferred to be done in integration test, but if integration test doesn't cover RBAC it needs to be tested in e2e.

**Comment by liggitt** [comment]
> Is RBAC unit test not enough? We can add an integration test or e2e test to have a DaemonSet with a ControllerRevision who doesn't match current DaemonSet template but has name collision with the ControllerRevision that the DaemonSet is going to create.
>
> It's preferred to be done in integration test, but if integration test doesn't cover RBAC it needs to be tested in e2e.

The unit test doesn't demonstrate that the permission is sufficient, it just records the fact that the permission was added to the role.

This type of externally visible behavior seems like an ideal candidate for the e2e suite (and possibly even a conformance test, to ensure that in a conformant cluster, the daemonset controller handles this case)

**Comment by kow3ns** [comment]
I don't think we should add ControllerRevision collision as an e2e test suite. Its not a black box test. I could see adding as an integration test though.

**Comment by liggitt** [comment]
> I don't think we should add ControllerRevision collision as an e2e test suite. The RBAC permissions associated with the DaemonSet controller are an implementation detail of the controller. The behavior with respect to the resolution of collisions for controller revisions is not visible to end users.

the net effect of the bug was definitely end-user visible (create/delete a few times hung the controller). That's a test gap I would want to ensure was covered in a actual cluster to flush out unexpected interactions (e.g. create, delete, create, delete, create, and ensure the expected pods materialize)

Phrasing it more in black box terms rather than trying to drive a hash collision makes sense to me. I think we'd want the scenario that broke to actually be tested, but if you want to merge this without addressing the test gap, I won't block on it.

/approve
policy change is approved

/hold
@kow3ns can unhold when satisfied with test coverage

**Comment by janetkuo** [comment]
I was not able to reproduce this by creating and deleting DaemonSets, and it's not clear to me how to reproduce it. Hash collision (i.e. ControllerRevision name collision) is different from what we saw in #62145, although it can trigger the same permission error. An e2e test for hash collision handling would have failed without this RBAC permission (it's more suitable to be an integration test, but integration test doesn't check RBAC), but it doesn't really cover the scenario described in this bug. 

OTOH this PR adds a missing RBAC permission that DaemonSet controller clearly needs.

**Comment by janetkuo** [comment]
Actually we should enable RBAC in DaemonSet integration tests instead of testing this in e2e just for RBAC.

**Comment by kow3ns** [comment]
/unhold
@janetkuo when you are satisfied I'm okay with merging this, can we open an issue to add RBAC to add the default RBAC for the workloads controller to the API server constructed for the integration framework.

**Comment by frodenas** [comment]
@janetkuo Thanks for taking a look at the tests, I haven't had time yet to look into this and probably I'll take me some time as I'm new to this codebase.

**Comment by janetkuo** [comment]
@frodenas thanks! Let's add the integration test in a follow-up PR

**Comment by janetkuo** [comment]
@frodenas would you cherrypick this fix to 1.10 also? Thanks!

**Comment by frodenas** [comment]
@janetkuo Done

**Pull request merged**
